File: Lesson 2 - Calculator (ObjC)/ObjCCalculator/ViewController.m
//
//  ViewController.m
//  ObjCCalculator
//
//  Created by Vince Li on 2016/7/28.
//  Copyright © 2016年 Vince. All rights reserved.
//

#import "ViewController.h"
#import "CalculatorBrain.h"

@interface ViewController ()

@property bool startToType;
@property NSMutableArray* touchHistoryArray;
@property (weak, nonatomic) IBOutlet UILabel *historyDisplay;
@property (weak, nonatomic) IBOutlet UILabel *display;
@property CalculatorBrain* brain;
@end

@implementation ViewController


- (void)viewDidLoad {
    [super viewDidLoad];
    _startToType = false;
    _brain = [[CalculatorBrain alloc] init];
}

- (IBAction)touchDigit:(UIButton *)sender {
    if([_brain isValidInput:sender.currentTitle currentDisplay:_display.text] == false) {
        return;
    }
    
    [_brain performOperarion:sender.currentTitle];
    
    if(_startToType == false) {
        _startToType = true;
        _display.text = sender.currentTitle;
    }
    else {
        _display.text = [_display.text stringByAppendingString: sender.currentTitle];
    }
    
    _historyDisplay.text = _brain.brainDescription;
}
- (IBAction)save {
    [_brain pushOperand:@"M"];
    [_brain.variableValues setObject:_display.text forKey:@"M"];
    _display.text = @([_brain evaluate]).stringValue;
}
- (IBAction)restore {
    NSString* text = [_brain.variableValues objectForKey:@"M"];
    if(text) {
        _display.text = text;
    }
    [_brain performOperarion:@"M"];
    
    _historyDisplay.text = _brain.brainDescription;
    _display.text = @(_brain.result).stringValue;
}

- (IBAction)performOperation:(UIButton *)sender {
    if([_brain isValidInput:sender.currentTitle currentDisplay:_display.text] == false) {
        return;
    }
    
    [_brain setOperand:_display.text];
    [_brain performOperarion:sender.currentTitle];
  
    _display.text = @(_brain.result).stringValue;
    _historyDisplay.text = _brain.brainDescription;
    if(_startToType == true) {
        _startToType = false;
    }
}

@end

File: Lesson 2 - Calculator (Swift)/SwiftCalculator/ViewController.swift
//
//  ViewController.swift
//  SwiftCalculator
//
//  Created by Vince Li on 2016/7/27.
//  Copyright © 2016年 Vince. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    
    private var startToType = false
    private var brain = CalculatorBrain()
    private var displayValue : Double? {
        get {
            if let value = display.text {
                return Double(value)
            }
            return nil
        }
        set {
            if let value = newValue {
                display.text = String(value)
            }
        }
    }
    @IBOutlet weak var touchHistory: UILabel!
    
    @IBAction func operationAction(sender: UIButton) {
        if let input = sender.currentTitle {
            if(brain.isValidInput(input, currentDisplay: display.text!) == false) {
                return;
            }
            
            brain.setOperand(displayValue)
            brain.performOperarion(input)
            
            displayValue = brain.result
            touchHistory.text = brain.description
            if(startToType) {
                startToType = false
            }
        }
    }
    
    @IBOutlet weak var display: UILabel!
    
    @IBAction func touchDigital(sender: UIButton) {
        if let input = sender.currentTitle {
            if(brain.isValidInput(input, currentDisplay: display.text!) == false) {
                return;
            }
            
            brain.performOperarion(input)
            
            if(startToType) {
                display.text = display.text! + input
            }
            else {
                display.text = input
                startToType = true
            }
            touchHistory.text = brain.description
        }
    }
    
    @IBAction func save() {
        brain.pushOperand("M")
        brain.variableValues["M"] = displayValue
        
        if let value = brain.evaluate() {
            displayValue = value
        }
    }
    @IBAction func restore() {
        displayValue = brain.variableValues["M"]
        brain.performOperarion("M")

        touchHistory.text = brain.description
        displayValue = brain.result
    }
}

